packages/react-dom/src/__tests__/ReactCompositeComponentState-test.js

```diff
@@ -500,18 +575,62 @@ describe('ReactCompositeComponent-state', () => {
       };
 
       const el = document.createElement('div');
-      ReactDOM.render(<Child incrementBy={0} />, el);
-      expect(el.textContent).toBe('count:1');
+      const root = ReactDOMClient.createRoot(el);
+      await act(() => {
+        root.render(<Child incrementBy={0} />);
+      });
 
-      ReactDOM.render(<Child incrementBy={2} />, el);
+      expect(el.textContent).toBe('count:1');
+      await act(() => {
+        root.render(<Child incrementBy={2} />);
+      });
       expect(el.textContent).toBe('count:3');
 
-      ReactDOM.render(<Child incrementBy={1} />, el);
+      await act(() => {
+        root.render(<Child incrementBy={1} />);
+      });
       expect(el.textContent).toBe('count:4');
     });
   }
 
-  it('should support setState in componentWillUnmount', () => {
+  it('should not support setState in componentWillUnmount', async () => {
+    let subscription;
+    class A extends React.Component {
+      componentWillUnmount() {
+        subscription();
+      }
+      render() {
+        return 'A';
+      }
+    }
+
+    class B extends React.Component {
+      state = {siblingUnmounted: false};
+      UNSAFE_componentWillMount() {
+        subscription = () => this.setState({siblingUnmounted: true});
+      }
+      render() {
+        return 'B' + (this.state.siblingUnmounted ? ' No Sibling' : '');
+      }
+    }
+
+    const el = document.createElement('div');
+    const root = ReactDOMClient.createRoot(el);
+    await act(() => {
+      root.render(<A />);
+    });
+    expect(el.textContent).toBe('A');
+
+    expect(() => {
+      ReactDOM.flushSync(() => {
+        root.render(<B />);
+      });
+    }).toErrorDev(
+      "Warning: Can't perform a React state update on a component that hasn't mounted yet",
+    );
+  });
+
+  it('Legacy mode should support setState in componentWillUnmount (#18851)', () => {
```

Comment:
Keeping this test which catches a behavior specific to `.render`. Added the test above which shows the error thrown in `createRoot` when trying to do this. Read the issue for more context.